tools/kth_threshold.cpp

```diff
@@ -0,0 +1,204 @@
+    mapper::map(index, m);
+
+    WandType wdata;
+
+    auto scorer = scorer::from_params(scorer_params, wdata);
+
+    mio::mmap_source md;
+    std::error_code error;
+    md.map(wand_data_filename, error);
+    if (error) {
+        spdlog::error("error mapping file: {}, exiting...", error.message());
+        std::abort();
+    }
+    mapper::map(wdata, md, mapper::map_flags::warmup);
+
+    using Pair = std::set<uint32_t>;
+    std::unordered_set<Pair, boost::hash<Pair>> pairs_set;
+
+    using Triple = std::set<uint32_t>;
+    std::unordered_set<Triple, boost::hash<Triple>> triples_set;
+
+    auto to_int = [](std::vector<std::string> const& term_ids) {
+        std::set<uint32_t> term_ids_int;
+        for (auto&& term_id: term_ids) {
+            try {
+                term_ids_int.insert(std::stoi(term_id));
+            } catch (const std::invalid_argument& e) {
+                throw std::runtime_error(fmt::format("Cannot convert {} to int.", term_id));
+            } catch (const std::out_of_range& e) {
+                throw std::runtime_error(fmt::format("Cannot convert {} to int.", term_id));
+            }
+        }
+        return term_ids_int;
+    };
+
+    std::string line;
+    if (pairs_filename) {
+        std::ifstream pin(*pairs_filename);
+        while (std::getline(pin, line)) {
+            std::vector<std::string> term_ids;
+            boost::algorithm::split(term_ids, line, boost::is_any_of(" \t"));
+            pairs_set.insert(to_int(term_ids));
+        }
+        spdlog::info("Number of pairs loaded: {}", pairs_set.size());
+    }
+
+    if (triples_filename) {
+        std::ifstream trin(*triples_filename);
+        while (std::getline(trin, line)) {
+            std::vector<std::string> term_ids;
+            boost::algorithm::split(term_ids, line, boost::is_any_of(" \t"));
+            triples_set.insert(to_int(term_ids));
+        }
+        spdlog::info("Number of triples loaded: {}", triples_set.size());
+    }
+
+    for (auto const& query: queries) {
+        float threshold = 0;
+
+        auto terms = query.terms;
+        topk_queue topk(k);
+        wand_query wand_q(topk);
+
+        for (auto&& term: terms) {
+            Query query;
+            query.terms.push_back(term);
+            wand_q(make_max_scored_cursors(index, wdata, *scorer, query), index.num_docs());
+            threshold = std::max(threshold, topk.size() == k ? topk.threshold() : 0.0f);
+        }
+        for (size_t i = 0; i < terms.size(); ++i) {
+            for (size_t j = i + 1; j < terms.size(); ++j) {
+                if (pairs_set.count({terms[i], terms[j]}) > 0) {
+                    Query query;
+                    query.terms.push_back(terms[i]);
+                    query.terms.push_back(terms[j]);
+                    wand_q(make_max_scored_cursors(index, wdata, *scorer, query), index.num_docs());
+                    threshold = std::max(threshold, topk.size() == k ? topk.threshold() : 0.0f);
```

Comment:
Just double-checking: are you sure that the function `size` does what it is supposed to and doesn't return the capacity instead? I remember something like that in the past but don't remember if was fixed.